src/server/pfs/server/driver.go

```diff
@@ -256,46 +251,65 @@ func (d *driver) inspectRepo(txnCtx *txncontext.TransactionContext, repo *pfs.Re
+func (d *driver) listRepo(ctx context.Context, includeAuth bool, repoType string, projectsFilter map[string]bool, cb func(*pfs.RepoInfo) error) error {
 	authSeemsActive := true
 	repoInfo := &pfs.RepoInfo{}
 
 	processFunc := func(string) error {
+		// Assume the user meant all projects by not providing any projects to filter on.
+		if len(projectsFilter) > 0 && !projectsFilter[repoInfo.Repo.Project.Name] {
+			return nil
+		}
 		size, err := d.repoSize(ctx, repoInfo.Repo)
 		if err != nil {
 			return err
 		}
 		repoInfo.SizeBytesUpperBound = size
-		if includeAuth && authSeemsActive {
-			permissions, roles, err := d.getPermissions(ctx, repoInfo.Repo)
-			if err == nil {
-				repoInfo.AuthInfo = &pfs.RepoAuthInfo{Permissions: permissions, Roles: roles}
-			} else if auth.IsErrNotActivated(err) {
-				authSeemsActive = false
+
+		// TODO cache the project check
+		// Check if the user can list repo within the project, otherwise check if they can read the repo.
+		if err := d.env.AuthServer.CheckProjectIsAuthorized(ctx, repoInfo.Repo.Project, auth.Permission_PROJECT_LIST_REPO); err != nil {
+			if errors.As(err, &auth.ErrNotAuthorized{}) {
+				if err := d.env.AuthServer.CheckRepoIsAuthorized(ctx, repoInfo.Repo, auth.Permission_REPO_READ); err != nil {
+					if errors.As(err, &auth.ErrNotAuthorized{}) {
+						return nil
+					}
+					return errors.Wrap(err, "could not check user is authorized to access repo")
+				}
+				// Here we know that the user does not have permission to list repos at the project level,
+				// but they should still see the repo because they have read permission.
 			} else {
-				return errors.Wrapf(grpcutil.ScrubGRPC(err), "error getting access level for %q", repoInfo.Repo)
+				return errors.Wrap(err, "could not check user is authorized to access project")
 			}
 		}
+		if authSeemsActive && includeAuth {
+			permissions, roles, err := d.getPermissions(ctx, repoInfo.Repo)
+			if err != nil {
+				if errors.Is(err, auth.ErrNotActivated) {
+					authSeemsActive = false
+					return cb(proto.Clone(repoInfo).(*pfs.RepoInfo))
+				} else {
+					return errors.Wrapf(err, "error getting access level for %q", repoInfo.Repo)
+				}
+			}
+			repoInfo.AuthInfo = &pfs.RepoAuthInfo{Permissions: permissions, Roles: roles}
+		}
 		return cb(proto.Clone(repoInfo).(*pfs.RepoInfo))
 	}
 
 	if repoType == "" {
 		// blank type means return all
-		return errors.EnsureStack(d.repos.ReadOnly(ctx).List(repoInfo, col.DefaultOptions(), processFunc))
-	} else {
-		err := d.repos.ReadOnly(ctx).GetByIndex(pfsdb.ReposTypeIndex, repoType, repoInfo, col.DefaultOptions(), processFunc)
-		return errors.EnsureStack(err)
+		return errors.Wrap(d.repos.ReadOnly(ctx).List(repoInfo, col.DefaultOptions(), processFunc), "could not list repos of all types")
 	}
+	return errors.Wrap(d.repos.ReadOnly(ctx).GetByIndex(pfsdb.ReposTypeIndex, repoType, repoInfo, col.DefaultOptions(), processFunc), "could not to list repos")
```

Comment:
`could not list repos`

Although: I prefer to let the caller say ‘could not list repos’; I would return something like ‘could not get repos of type spec: ERROR FROM GetByIndex’. That way each function returns what went wrong from its point of view.